From cbd13f6ce6601387ea59807d899ac378c8fd39b8 Mon Sep 17 00:00:00 2001
From: Pranav Kant <pranav913@gmail.com>
Date: Mon, 6 Oct 2014 16:51:58 +0200
Subject: [PATCH 1/3] embed: Switch to the preview as soon as an item starts
 loading

We are essentially in the preview, not the overview, when an item
starts loading. Therefore the UI should relate to the current context.

https://bugzilla.gnome.org/show_bug.cgi?id=720516
---
 src/embed.js | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/embed.js b/src/embed.js
index 8d69b78..0bcdcef 100644
--- a/src/embed.js
+++ b/src/embed.js
@@ -384,6 +384,8 @@ const Embed = new Lang.Class({
     },
 
     _onLoadStarted: function() {
+        Application.modeController.setWindowMode(WindowMode.WindowMode.PREVIEW);
+
         this._clearLoadTimer();
         this._loadShowId = Mainloop.timeout_add(_PDF_LOADER_TIMEOUT, Lang.bind(this,
             function() {
@@ -396,8 +398,6 @@ const Embed = new Lang.Class({
     },
 
     _onLoadFinished: function(manager, doc, docModel) {
-        Application.modeController.setWindowMode(WindowMode.WindowMode.PREVIEW);
-
         docModel.set_sizing_mode(EvView.SizingMode.AUTOMATIC);
         docModel.set_page_layout(EvView.PageLayout.AUTOMATIC);
         this._toolbar.setModel(docModel);
@@ -410,8 +410,6 @@ const Embed = new Lang.Class({
     },
 
     _onLoadError: function(manager, doc, message, exception) {
-        Application.modeController.setWindowMode(WindowMode.WindowMode.PREVIEW);
-
         this._clearLoadTimer();
         this._spinnerBox.stop();
         this._setError(message, exception.message);
@@ -455,6 +453,8 @@ const Embed = new Lang.Class({
         // pack the toolbar
         this._toolbar = new Preview.PreviewToolbar(this._preview);
         this._titlebar.add(this._toolbar.widget);
+
+        this._stack.set_visible_child_name('preview');
     },
 
     _prepareForEdit: function() {
-- 
2.1.0


From 95c0f964979882b9e884f15576a38a2080f9bcd0 Mon Sep 17 00:00:00 2001
From: Debarshi Ray <debarshir@gnome.org>
Date: Mon, 6 Oct 2014 18:13:44 +0200
Subject: [PATCH 2/3] Don't show a glimpse of the older item when going back to
 the preview

Now that we switch to the preview as soon as an item starts loading, we
should take care to clear any older item that might have been previewed
earlier.

Since EvView does not allow unsetting the model (by passing NULL), we
destroy the old widget and create a new one. We do the same for the
NavControls because of GdNavBar.

https://bugzilla.gnome.org/show_bug.cgi?id=720516
---
 src/embed.js   |  2 +-
 src/preview.js | 18 +++++++++++++++---
 2 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/src/embed.js b/src/embed.js
index 0bcdcef..d390d0f 100644
--- a/src/embed.js
+++ b/src/embed.js
@@ -430,7 +430,7 @@ const Embed = new Lang.Class({
 
     _prepareForOverview: function() {
         if (this._preview)
-            this._preview.setModel(null);
+            this._preview.reset();
         if (this._edit)
             this._edit.setUri(null);
         if (this._toolbar)
diff --git a/src/preview.js b/src/preview.js
index 9431b8b..878925c 100644
--- a/src/preview.js
+++ b/src/preview.js
@@ -80,9 +80,6 @@ const PreviewView = new Lang.Class({
         this._previewContextMenu = Gtk.Menu.new_from_model(model);
         this._previewContextMenu.attach_to_widget(this.widget, null);
 
-        // create page nav controls
-        this._navControls = new PreviewNavControls(this, this._overlay);
-
         this.widget.show_all();
 
         this._bookmarkPage = Application.application.lookup_action('bookmark-page');
@@ -339,6 +336,8 @@ const PreviewView = new Lang.Class({
             this._onViewSelectionChanged));
         this.view.connect('external-link', Lang.bind(this,
             this._handleExternalLink));
+
+        this._navControls = new PreviewNavControls(this, this._overlay);
     },
 
     _getPreviewContextMenu: function() {
@@ -504,6 +503,13 @@ const PreviewView = new Lang.Class({
         this.emit('search-changed', job.has_results());
     },
 
+    reset: function() {
+        this.setModel(null);
+        this.view.destroy();
+        this._navControls.destroy();
+        this._createView();
+    },
+
     setModel: function(model) {
         if (this._model == model)
             return;
@@ -740,6 +746,12 @@ const PreviewNavControls = new Lang.Class({
     hide: function() {
         this._visible = false;
         this._updateVisibility();
+    },
+
+    destroy: function() {
+        this.bar_widget.destroy();
+        this.prev_widget.destroy();
+        this.next_widget.destroy();
     }
 });
 
-- 
2.1.0


From 63aa67d248b7bcddc0025222216ee83ee8fbdbe8 Mon Sep 17 00:00:00 2001
From: Debarshi Ray <debarshir@gnome.org>
Date: Tue, 7 Oct 2014 15:04:58 +0200
Subject: [PATCH 3/3] application, preview: Disable the gear menu while a
 document is loading

https://bugzilla.gnome.org/show_bug.cgi?id=720516
---
 src/application.js | 3 +--
 src/preview.js     | 3 ++-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/application.js b/src/application.js
index e31c643..e6aad0f 100644
--- a/src/application.js
+++ b/src/application.js
@@ -467,8 +467,7 @@ const Application = new Lang.Class({
             { name: 'gear-menu',
               callback: this._onActionToggle,
               state: GLib.Variant.new('b', false),
-              accel: 'F10',
-              window_mode: WindowMode.WindowMode.PREVIEW },
+              accel: 'F10' },
             { name: 'view-as',
               callback: this._onActionViewAs,
               create_hook: this._viewAsCreateHook,
diff --git a/src/preview.js b/src/preview.js
index 878925c..f1e49cd 100644
--- a/src/preview.js
+++ b/src/preview.js
@@ -772,7 +772,7 @@ const PreviewToolbar = new Lang.Class({
         this._searchAction.enabled = false;
 
         this._gearMenu = Application.application.lookup_action('gear-menu');
-        this._gearMenu.enabled = true;
+        this._gearMenu.enabled = false;
 
         // back button, on the left of the toolbar
         let backButton = this.addBackButton();
@@ -857,6 +857,7 @@ const PreviewToolbar = new Lang.Class({
             return;
 
         this._model = model;
+        this._gearMenu.enabled = true;
         this._enableSearch();
         this._setToolbarTitle();
     }
-- 
2.1.0

